lib/ukallocbuddy: Add debug freelist sanity checks

andreittr opened the pull request:
### Description of changes

This adds a Kconfig option to enable sanity checking of buddy allocator free lists at runtime. Check points are placed at the beginning and end of functions that operate on the free lists. When the option is disabled the checks default to zero-overhead no-ops.
In addition, this adds an assert that returned allocated pages are correctly aligned to their buddy order.

Furthermore, this PR preemptively hardens the page allocator against pathologically large numbers of requested pages which might cause out-of-bounds access to the freelist head array.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

Can be tested with the code snippet from https://github.com/unikraft/unikraft/pull/931 (without applying that patch), should complain after the `free()` call.

andreittr committed:
lib/ukallocbuddy: Add debug freelist sanity checks

This adds a Kconfig option to enable sanity checking of buddy allocator
free lists at runtime. Check points are placed at the beginning and end
of functions that operate on the free lists. When the option is disabled
the checks default to zero-overhead no-ops.
In addition, this adds an assert that returned allocated pages are
correctly aligned to their buddy order.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

Starnox reviewed (approved):
It seems to catch the errors I tried throwing at it. Performance is not impacted when the option is not selected. For me it looks like a nice addition.

Reviewed-by: Eduard-Florin Mihailescu <mihailescu.eduard@gmail.com>

andreittr committed:
lib/ukallocbuddy: Harden buddy allocator OOM check

When the buddy allocator chooses a page order to allocate it can (for
pathologically large values of num_pages) start from a value larger than
FREELIST_SIZE, leading to out-of-bounds access of the freelist array.
This change hardens the out-of-memory check to prevent this.

Signed-off-by: Andrei Tatar <andrei@unikraft.io>

andreittr commented:
Update: additional commit hardening the out-of-memory check in the buddy page allocator.

Starnox reviewed (approved):
LGTM!

Reviewed-by: Eduard-Florin Mihailescu <mihailescu.eduard@gmail.com>

andreittr commented:
Rebased on staging & fixed some style issues (remaining checkpatch failures expected; will propose fixes to checkpatch).

StefanJum reviewed (approved):
Thank you @andreittr, all good.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
